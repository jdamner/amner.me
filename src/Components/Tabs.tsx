'use client'
import React from "react";
/* API */
import { useState } from "react";

/* Components */
import TabButton from "../Atoms/Links/TabButton";
import ReactMarkdown from 'react-markdown';

export default function Tabs({ tabs, defaultContent, title }:
  {
    tabs: {
      title: string,
      content: string
    }[],
    defaultContent: string,
    title: React.ReactNode
  }
) {

  const [activeTab, setActiveTab] = useState(0);
  const [tabOpen, setTabOpen] = useState(false);

  const handleTabChange = (index: number) => {
    setTabOpen(index === activeTab ? !tabOpen : true);
    setActiveTab(index);
  }

  const TabButtons = tabs.map((tab, index) => {
    const active = tabOpen && activeTab === index || !tabOpen;
    const ariaCurrent = activeTab === index && tabOpen ? 'page' : undefined;
    return (
      <TabButton
        key={index}
        delay={index * 0.25}
        role="tab"
        id={`tab-${index}`}
        aria-current={ariaCurrent}
        aria-controls={`tab-${index}`}
        active={active}
        onClick={() => handleTabChange(index)}
      >{tab.title}</TabButton>
    );
  });

  const contentClassName = 'prose prose-slate dark:prose-invert mr-auto text-slate-900 dark:text-slate-200'

  const ActiveTab = () => {
    const tab = tabs[activeTab];
    return (
      <div
        className={contentClassName}
        id={`tab-${activeTab}`}
        role="tabpanel"
        aria-labelledby={`tab-${activeTab}`}
      >
        <ReactMarkdown
          components={
            { a: (props) => <a {...props} target="_blank" rel="noopener noreferrer" /> }
          }>{tab.content}</ReactMarkdown>
      </div>
    )
  }


  const NonActiveTab = () => {
    return (
      <ReactMarkdown className={contentClassName}>{defaultContent}</ReactMarkdown>
    )
  }

  return (
    <>
      {title}
      <div className="md:grid md:grid-cols-3" role="tablist">
        <div className="flex flex-row md:flex-col flex-wrap md:flex-nowrap gap-3 md:gap-0 items-start md:items-end pr-5 mb-3">
          {TabButtons}
        </div>
        <div className="md:col-span-2">
          {tabOpen ? <ActiveTab /> : <NonActiveTab />}
        </div>
      </div>
    </>
  )
}
