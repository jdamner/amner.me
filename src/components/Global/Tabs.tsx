import { useState } from "react";
import { event } from "../../api/Insights";

import type { post } from "../../types/post.type";

import TabButton from "../Links/TabButton";
import Title from "./Title";
import ReactMarkdown from 'react-markdown';


export default function Tabs({ tabs, content }: { tabs: post[], content: string }): JSX.Element {

  const [activeTab, setActiveTab] = useState(0);
  const [tabOpen, setTabOpen] = useState(false);

  const handleTabChange = (index) => {
    event('tab-change', { index, tabOpen });
    setTabOpen(index === activeTab ? !tabOpen : true);
    setActiveTab(index);
  }

  const TabButtons = tabs.map((tab, index) => {
    const active = tabOpen && activeTab === index || !tabOpen;
    return (
      <TabButton
        key={index}
        index={index + 1}
        name={tab.title}
        active={active}
        onClick={() => handleTabChange(index)}
      />
    );
  });

  const contentClassName = 'prose prose-slate dark:prose-invert mr-auto px-3 md:px-0 text-slate-900 dark:text-slate-200'

  const ActiveTab = () => {
    const tab = tabs[activeTab];
    return (
      <ReactMarkdown className={contentClassName}>{tab.content}</ReactMarkdown>
    )
  }


  const NonActiveTab = () => {
    return (
      <ReactMarkdown className={contentClassName}>{ content }</ReactMarkdown>
    )
  }

  return (
    <section className="container mx-auto my-5 py-5 px-3 md:px-0" id="services-tabs">
      <Title title={'Techincal Skills'}>Web Development</Title>
      <div className="md:grid md:grid-cols-3">
        <div className="flex flex-row md:flex-col flex-wrap md:flex-nowrap gap-3 md:gap-0 items-start md:items-end pr-5 mb-3">
          {TabButtons}
        </div>
        <div className="md:col-span-2">
          {tabOpen ? <ActiveTab /> : <NonActiveTab />}
        </div>
      </div>
    </section>
  )
}
