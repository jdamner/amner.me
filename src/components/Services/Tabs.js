import React from "react";
import Seperator from "../Global/Seperator";
import TabButton from "./TabButton";
import TabContent from "./TabContent";
import { StaticQuery, graphql } from "gatsby";
import { GatsbyImage, getImage } from "gatsby-plugin-image";

export default function Tabs() {

    const [activeTab, setActiveTab] = React.useState(0);
    const [tabOpen, setTabOpen] = React.useState(false);

    const handleTabChange = (index) => {
        setTabOpen(index === activeTab ? !tabOpen : true);
        setActiveTab(index);
    }

    return (
        <StaticQuery
            query={graphql`
            query MyQuery {
                allFile(filter: {sourceInstanceName: {eq: "services"}}) {
                  edges {
                    node {
                      id
                      childrenMarkdownRemark {
                        frontmatter {
                          title
                          thumbnail {
                            childImageSharp {
                              gatsbyImageData(
                                width: 200
                                placeholder: BLURRED
                                formats: [AUTO, WEBP, AVIF]
                              )
                            }
                          }
                        }
                        html
                      }
                    }
                  }
                }
              }`}
            render={data => {
                const { allFile } = data;
                const TabButtons = allFile.edges.map(edge => {

                    const { frontmatter } = edge.node.childrenMarkdownRemark[0];
                    if (!frontmatter) {
                        return null;
                    }

                    const file = frontmatter.thumbnail ? getImage(frontmatter.thumbnail) : null;
                    console.log(file, frontmatter);
                    const image = file ? <GatsbyImage image={file} alt={frontmatter.title} objectFit={'contain'} /> : null;
                    

                    return (
                        <TabButton
                            key={edge.node.id}
                            name={frontmatter.title}
                            image={image}
                            active={tabOpen && activeTab === edge.node.id}
                            index={edge.node.id}
                            onClick={() => handleTabChange(edge.node.id)}
                        />
                    );
                })

                const active = allFile.edges.find(edge => edge.node.id === activeTab)
                const tab = active ? active.node.childrenMarkdownRemark[0] : null;


                return (
                    <div className="tab">
                        <div className='blog-wrap-title'>
                            <Seperator />
                            <h2>Special Skills</h2>
                            <Seperator modifier='rotate' />
                        </div>
                        <div className="tab-buttons-wrapper">
                            {TabButtons}
                        </div>
                        <div className="tab-content-wrapper">
                            {tabOpen ? <TabContent tab={tab} /> : null}
                        </div>
                    </div>
                )
            }
            }
        />
    )
}