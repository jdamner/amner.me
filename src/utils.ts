/**
 * Helper functions
 */
import fs from 'fs'
import matter from 'gray-matter'
import { join } from 'path'
import { slug } from 'github-slugger'

import type { ReactNode } from 'react'
import type { MdFile } from './types'

/**
 * Makes an object JSON parseable
 * 
 * @param {any} obj The object to be parsed
 * @returns {any}
 */
export function makeJsonParseable<Type>(obj: Type): Type {
	return JSON.parse(JSON.stringify(obj)) as Type
}

/**
 * Reads a markdown file
 */
export async function readMdFile(filename: string, directory: string = 'content/'): Promise<MdFile> {
	const path = `${directory}/${filename}`
	const fullPath = `${process.cwd()}/${path}`
	const fileContents = fs.readFileSync(fullPath, 'utf8')
	return {
		slug: slug(filename.replace(/\.md$/, '')),
		...matter(fileContents),
	}
}

/**
 * Reads all slugs from a directory
 * 
 * @param {string} directory The directory to read from
 * 
 * @returns {string[]}
 */
export function getFilesInDir(directory: string): string[] {
	const fullPath = join(process.cwd(), directory)
	return fs.readdirSync(fullPath)
}

/**
 * Converts all headings in a markdown file to have IDs 
 * Converts 
 * ## Heading 1 
 * to become
 * ## Heading 1{#heading-1}
 * 
 * @param {string} markdown The markdown to convert
 * 
 * @returns {string}
 */
export function convertHeadingsToAnchorLinks(markdown: string): string {
	return markdown.replace(/(#{1,6})\s(.+)\n/gm, (match, p1, p2) => {
		return `${p1} ${p2}{#${slug(p2)}}\n`
	})
}


/**
 * Converts a react node to a slug
 * 
 * Usually the value is a string, but sometimes it contains
 * a `<code>` element, which is a react node. This function
 * converts the value to a string.
 */
export function makeSlug(value: ReactNode | ReactNode[]): string {
	if (! Array.isArray(value)) return slug(value?.toString() ?? '');
	return slug(
		value.map((node) => {
			if (typeof node === 'string' ||
				typeof node === 'number' ||
				typeof node === 'boolean'
			) return node;

			if (typeof node === 'object' && node !== null && 'props' in node) {
				return makeSlug(node.props.children)
			}

			return ''
		}).join('')
	)
}