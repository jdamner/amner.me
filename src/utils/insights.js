import { track, parameters, init } from 'insights-js';

const pageView = () => {
    event('page-view');
}

const event = (eventName, eventProperties = {}) => {
    init('HbSIygPWVvMzXFRP');
    track({
        id: eventName,
        parameters: {
            ...parameters,
            ...eventProperties,
        },
    });
}

export { pageView, event };