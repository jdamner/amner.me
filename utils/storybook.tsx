import { faker } from "@faker-js/faker";
import { MdFile, PostLinkType } from "../types";
import matter from "gray-matter";
import React from "react";

type MdAttributes = Record<string, string>;

faker.seed(123);

export const generateMdFile = <T extends MdAttributes>(
  attributes: T,
  content?: string,
): Omit<MdFile<string>, "data"> & { data: T } => {
  return matter(generateMD(attributes, content)) as Omit<
    MdFile<string>,
    "data"
  > & { data: T };
};

export const generateMdContent = (): string => {
  return (
    `# ${faker.lorem.words(3)}\n` +
    Array.from({ length: 3 }, () => {
      return (
        `## ${faker.lorem.words(2)}\n` +
        faker.lorem.paragraph() +
        "\n" +
        Array.from({ length: 3 }, () => {
          return (
            `### ${faker.lorem.words(2)}\n` + `${faker.lorem.paragraph()}\n`
          );
        }).join("")
      );
    }).join("")
  );
};

export const generatePostLink = (): PostLinkType => {
  return {
    title: faker.lorem.words(3),
    thumbnail: faker.image.url(),
    slug: faker.lorem.slug(),
  };
};

export const generatePostLinks = (count: number): PostLinkType[] => {
  return Array.from({ length: count }, () => generatePostLink());
};

export const generateMD = (
  attributes: MdAttributes,
  content?: string,
): string => {
  const frontMatter = generateMDFrontMatter(attributes);
  return `${frontMatter}\n${content ?? generateMdContent()}`;
};

export const generateMDFrontMatter = (attributes: MdAttributes): string => {
  return (
    "---\n" +
    Object.keys(attributes)
      .map((key) => `${key}: ${attributes[key]}`)
      .join("\n") +
    "\n---"
  );
};

export const generatePlaceholderContent = (
  size: number = 5,
): React.ReactNode => {
  return (
    <>
      {Array.from({ length: size }).map((_, i) => (
        <p key={i}>{faker.lorem.paragraph(i + 1)}</p>
      ))}
    </>
  );
};
