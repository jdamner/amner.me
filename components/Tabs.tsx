"use client";

import React from "react";
/* API */
import { useState } from "react";

/* Components */
import TabButton from "./Atoms/Links/TabButton";

import { useAllServices } from "../utils";

export default function Tabs({
  tabs,
  defaultContent,
  title,
}: {
  tabs: ReturnType<typeof useAllServices>;
  defaultContent: React.ReactNode;
  title: React.ReactNode;
}) {
  const [activeTab, setActiveTab] = useState(0);
  const [tabOpen, setTabOpen] = useState(false);

  const handleTabChange = (index: number) => {
    setTabOpen(index === activeTab ? !tabOpen : true);
    setActiveTab(index);
  };

  const TabButtons = tabs.map((tab, index) => {
    const active = tabOpen && activeTab === index;
    const ariaCurrent = activeTab === index && tabOpen ? "page" : undefined;
    return (
      <TabButton
        key={index}
        delay={index * 0.25}
        role="tab"
        id={`tab-${index}`}
        aria-current={ariaCurrent}
        aria-controls={`tab-${index}`}
        active={active}
        onClick={() => handleTabChange(index)}
      >
        {tab.data.title}
      </TabButton>
    );
  });

  const contentClassName =
    "prose prose-slate dark:prose-invert mr-auto text-slate-900 dark:text-slate-200";

  function ActiveTab() {
    const tab = tabs[activeTab];
    return (
      <div
        className={contentClassName}
        id={`tab-${activeTab}`}
        role="tabpanel"
        aria-labelledby={`tab-${activeTab}`}
      >
        {tab.content}
      </div>
    );
  }

  function NonActiveTab() {
    return <div className={contentClassName}>{defaultContent}</div>;
  }

  return (
    <>
      {title}
      <div className="md:grid md:grid-cols-3" role="tablist">
        <div className="flex flex-row md:flex-col flex-wrap md:flex-nowrap gap-3 md:gap-0 items-start md:items-end pr-5 mb-3">
          {TabButtons}
        </div>
        <div className="md:col-span-2">
          {tabOpen ? <ActiveTab /> : <NonActiveTab />}
        </div>
      </div>
    </>
  );
}
