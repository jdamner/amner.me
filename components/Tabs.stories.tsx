import { faker } from "@faker-js/faker";
import type { Meta, StoryObj } from "@storybook/react";
import { generateMdFile } from "@/utils/storybook";

import Tabs from "./Tabs";

type Story = StoryObj<typeof Tabs>;

export default {
  title: "Components/Tabs",
  component: Tabs,
} as Meta;

faker.seed(123);

export const Default: Story = {
  args: {
    title: "",
    defaultContent: faker.lorem.paragraphs(2),
    tabs: Array.from({ length: 5 }, (_, i) =>
      generateMdFile({ title: `Tab ${i + 1}` }),
    ),
  },
};
