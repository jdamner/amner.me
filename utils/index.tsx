/* External Deps */
import { readFileSync, readdirSync } from "fs";
import { resolve } from "path";
import matter from "gray-matter";
import { slug } from "github-slugger";

/* Components */
import { Markdown } from "@/components/Markdown";

/* Constants */
const postsDirectory: string = resolve("content/blog/");
const servicesDirectory: string = resolve("content/services");
const cvDirectory: string = resolve("content/cv");
const projectsDirectory: string = resolve(cvDirectory, "projects");
const referencesDirectory: string = resolve(cvDirectory, "references");
const employmentDirectory: string = resolve(cvDirectory, "employment");
const educationDirectory: string = resolve(cvDirectory, "education");

type MatterData = ReturnType<typeof matter>["data"];
type AttributeValidator<T extends MatterData> = (data: MatterData) => data is T;

const defaultAttributeValidator = <T extends MatterData>(
  data: MatterData,
): data is T => {
  return typeof data === "object" && data !== null;
};

/**
 * Gets all files in a directory
 */
const getAllFrom = <T extends MatterData>(
  directory: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  readdirSync(directory).map((filename) =>
    readMdFile<T>(filename, directory, attributeValidator),
  );

/**
 * Reads a markdown file
 */
const readMdFile = <T extends MatterData>(
  filename: string,
  directory: string = "content/",
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  parseMarkdown<T>(
    readFileSync(`${directory}/${filename}`, "utf8"),
    filename,
    attributeValidator,
  );

/**
 * Parses markdown file contents
 */
export const parseMarkdown = <T extends MatterData>(
  fileContents: string,
  filename: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) => {
  const { content, data } = matter(fileContents);

  if (!attributeValidator(data)) {
    console.log(data);
    throw new Error(`Invalid attributes in ${filename}` + JSON.stringify(data));
  }

  return {
    data,
    slug: slug(filename.replace(/\.mdx$/, "")),
    content: <Markdown>{content}</Markdown>,
  };
};

/**
 * Gets all Posts
 */
const getAllPosts = () =>
  getAllFrom<WithTitle & WithThumbnail & WithDate>(
    postsDirectory,
    (data): data is WithTitle & WithThumbnail & WithDate =>
      hasTitle(data) && hasThumbnail(data) && hasDate(data),
  );

export const useAllServices = () =>
  getAllFrom<WithTitle>(servicesDirectory, hasTitle);

export const getAllProjects = () =>
  getAllFrom<WithTitle & WithLink>(
    projectsDirectory,
    (data): data is WithTitle & WithLink => hasTitle(data) && hasLink(data),
  );

export const useAllReferences = () =>
  getAllFrom<WithTitle & WithWebsite & WithPhone & WithEmail>(
    referencesDirectory,
    (data): data is WithTitle & WithWebsite & WithPhone & WithEmail =>
      hasTitle(data) && hasWebsite(data) && hasPhone(data) && hasEmail(data),
  );

export const useAllEmployment = () =>
  getAllFrom<
    WithTitle & WithSubtitle & WithDateString & WithRole & WithWebsite
  >(
    employmentDirectory,
    (
      data,
    ): data is WithTitle &
      WithSubtitle &
      WithDateString &
      WithRole &
      WithWebsite =>
      hasTitle(data) &&
      hasSubtitle(data) &&
      hasDateString(data) &&
      hasRole(data) &&
      hasWebsite(data),
  );

export const useAllEducation = () =>
  getAllFrom<WithTitle & WithDateString>(
    educationDirectory,
    (data): data is WithTitle & WithDateString =>
      hasTitle(data) && hasDateString(data),
  );

export const getPost = (slug: string) =>
  getAllPosts().find((post) => post.slug === slug);

export const getAllPostLinks = () =>
  getAllPosts()
    .map((post) => ({
      title: post.data.title,
      thumbnail: post.data.thumbnail,
      slug: post.slug,
      date: post.data.date,
    }))
    .sort((a, b) => (a.date > b.date ? -1 : 1));

type WithTitle = { title: string };
type WithSubtitle = { subtitle: string };
type WithLink = { link: string };
type WithThumbnail = { thumbnail: string };
type WithDate = { date: Date };
type WithDateString = { date: string };
type WithPhone = { phone: string };
type WithWebsite = { website: string };
type WithEmail = { email: string };
type WithRole = { role: string };

const hasTitle = (data: MatterData): data is WithTitle =>
  defaultAttributeValidator(data) &&
  "title" in data &&
  typeof data.title === "string";

const hasSubtitle = (data: MatterData): data is WithSubtitle =>
  "subtitle" in data && typeof data.subtitle === "string";

const hasLink = (data: MatterData): data is WithLink =>
  "link" in data && typeof data.link === "string";

const hasThumbnail = (data: MatterData): data is WithThumbnail =>
  "thumbnail" in data && typeof data.thumbnail === "string";

const hasDate = (data: MatterData): data is WithDate =>
  "date" in data && data.date instanceof Date;

const hasDateString = (data: MatterData): data is WithDateString =>
  "date" in data && typeof data.date === "string";

const hasPhone = (data: MatterData): data is WithPhone =>
  "phone" in data && typeof data.phone === "string";

const hasWebsite = (data: MatterData): data is WithWebsite =>
  "website" in data && typeof data.website === "string";

const hasEmail = (data: MatterData): data is WithEmail =>
  "email" in data && typeof data.email === "string";

export const hasRole = (data: MatterData): data is WithRole =>
  "role" in data && typeof data.role === "string";
