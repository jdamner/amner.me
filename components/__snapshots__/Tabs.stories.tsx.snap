// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Tabs Default smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="md:grid md:grid-cols-3"
       role="tablist"
  >
    <div class="flex flex-row md:flex-col flex-wrap md:flex-nowrap gap-3 md:gap-0 items-start md:items-end pr-5 mb-3">
      <button delay="0"
              role="tab"
              id="tab-0"
              aria-controls="tab-0"
              class="undefined text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline text-gray-600 dark:text-gray-400"
      >
        Tab 1
      </button>
      <button delay="0.25"
              role="tab"
              id="tab-1"
              aria-controls="tab-1"
              class="undefined text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline text-gray-600 dark:text-gray-400"
      >
        Tab 2
      </button>
      <button delay="0.5"
              role="tab"
              id="tab-2"
              aria-controls="tab-2"
              class="undefined text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline text-gray-600 dark:text-gray-400"
      >
        Tab 3
      </button>
      <button delay="0.75"
              role="tab"
              id="tab-3"
              aria-controls="tab-3"
              class="undefined text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline text-gray-600 dark:text-gray-400"
      >
        Tab 4
      </button>
      <button delay="1"
              role="tab"
              id="tab-4"
              aria-controls="tab-4"
              class="undefined text-xl md:text-3xl font-bold mr-3 md:mr-0 group hover:underline text-gray-600 dark:text-gray-400"
      >
        Tab 5
      </button>
    </div>
    <div class="md:col-span-2">
      <div class="prose prose-slate dark:prose-invert mr-auto text-slate-900 dark:text-slate-200">
        Summisse circumvenio curto calamitas studio ipsa super super. Curo termes vomer cunctatio stillicidium nisi. Apud cruciamentum cui conicio quis suspendo.
Decens caries adulatio sub cubicularis. Taceo ratione auctor decerno atque agnosco exercitationem. Exercitationem cursus reiciendis coadunatio ulciscor deprimo surculus taedium.
      </div>
    </div>
  </div>
</div>
`;
